src/lib/ast_traversal.rs

```diff
@@ -0,0 +1,428 @@
+use logos::Span;
+
+use crate::ast::{
+    Case, Dec, DecField, Exp, ExpField, Loc, Node, Pat, PatField, Source, Type, TypeBind, TypeField,
+};
+
+// TODO: move to another file
+pub fn get_breakpoint_span_from_line<'a>(
+    tree: &'a Loc<SyntaxTree<'a>>,
+    breakpoint_line: usize,
+) -> Option<Span> {
+    match &tree.1 {
+        Source::Known { span, line, .. } => {
+            if *line == breakpoint_line {
+                Some(span.clone())
+            } else if *line < breakpoint_line {
+                let mut span = None;
+                tree.for_each_child(|x| {
+                    match (&span, get_breakpoint_span_from_line(x, breakpoint_line)) {
+                        (None, Some(s)) => span = Some(s),
+                        _ => {}
+                    }
+                });
+                span
+            } else {
+                None
+            }
+        }
+        _ => None,
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SyntaxTree<'a> {
+    Exp(&'a Exp),
+    Dec(&'a Dec),
+    Pat(&'a Pat),
+    Type(&'a Type),
+    ExpField(&'a ExpField),
+    DecField(&'a DecField),
+    PatField(&'a PatField),
+    TypeField(&'a TypeField),
+    Case(&'a Case),
+    TypeBind(&'a TypeBind),
+}
+
+pub trait ToTree<'a> {
+    fn tree(&'a self) -> Loc<SyntaxTree<'a>>;
+}
+
+impl<'a, S: ToNode> ToTree<'a> for Node<S> {
+    fn tree(&'a self) -> Loc<SyntaxTree<'a>> {
+        S::node_tree(self)
+    }
+}
+
+impl<'a> ToTree<'a> for Loc<SyntaxTree<'a>> {
+    fn tree(&'a self) -> Loc<SyntaxTree<'a>> {
+        Loc(self.0.clone(), self.1.clone()) // TODO pointer instead of clone?
```

Comment:
We may want `Rc` in place of `Box` in our AST definitions, since in general I think we want to do O(1) clones of the ast to get around borrow checking constraints.  So that is to say, I think cloning is okay, and sometimes unavoidable.